README.md
# B&eacute;zier interpolation

This module calculates a set of B&eacute;zier splines that represent a continuous curve
through a set of points.  Both integer and floating point versions are provided.
I use this to create curves in SVG files.

The algorithm was ported from the Javascript on
[this page](https://www.particleincell.com/wp-content/uploads/2012/06/bezier-spline.js).
The page describing what the code does is
[here](https://www.particleincell.com/2012/bezier-splines/).  This page also provides
a little interactive demo.

If you pass in N + 1 knots (points to be
interpolated), you will get N sets of control points back, one set for each part of
the curve.  Each set has four control points.  The first control point (P0) is the
starting knot, and the last control point (P3) is the ending knot.  P1 and P2 are
determined by the module.  A sample image is shown below, with the knots in blue and
the computed control points in green.

![sample SVG](sample.svg)


CHANGELOG.md
# Change Log

## [1.0.1] - 2022-07-21

Corrected name on license.

## [1.0.0] - 2022-07-11

Initial release.


catmullrom.go
package bezier

import (
  "image"
)

// The "F" suffix on the struct names means float64.

type PointF struct {
  X, Y float64
}

type BezierF struct {
  P0, P1, P2, P3 PointF
}

type Bezier struct {
  P0, P1, P2, P3 image.Point
}

func GetControlPointsI(p []image.Point) []Bezier {
  // Convert integer values to floats, calcalute control points, and then
  // convert back to integers.
  pf := make([]PointF, len(p), len(p))
  for j, source := range(p) {
    pf[j].X = float64(source.X)
    pf[j].Y = float64(source.Y)
  }
  bezierF := GetControlPointsF(pf)
  bezier := make([]Bezier, len(bezierF), len(bezierF))
  for j, bz := range(bezierF) {
    bezier[j].P0.X = int(bz.P0.X)
    bezier[j].P0.Y = int(bz.P0.Y)
    bezier[j].P1.X = int(bz.P1.X)
    bezier[j].P1.Y = int(bz.P1.Y)
    bezier[j].P2.X = int(bz.P2.X)
    bezier[j].P2.Y = int(bz.P2.Y)
    bezier[j].P3.X = int(bz.P3.X)
    bezier[j].P3.Y = int(bz.P3.Y)
  }
  return bezier
}

// Algorithm ported from https://www.particleincell.com/wp-content/uploads/2012/06/bezier-spline.js
func GetControlPointsF(p []PointF) []BezierF {
  n := len(p) - 1 // number of segments
  bezier := make([]BezierF, n, n)

  // Fill in the end points of the Bezier curves.
  for j, bz := range(bezier) {
    bz.P0 = p[j]
    bz.P3 = p[j+1]
  }

  // Slices needed for calculations, listed as "rhs vector" on page listed above.
  a := make([]PointF, n, n)
  b := make([]PointF, n, n)
  c := make([]PointF, n, n)
  r := make([]PointF, n, n)

  a[0].X, a[0].Y = 0.0, 0.0
  b[0].X, b[0].Y = 2.0, 2.0
  c[0].X, c[0].Y = 1.0, 1.0
  r[0].X = p[0].X + 2.0 * p[1].X
  r[0].Y = p[0].Y + 2.0 * p[1].Y

  for j := 1; j < (n-1); j++ {
    a[j].X, a[j].Y = 1.0, 1.0
    b[j].X, b[j].Y = 4.0, 4.0
    c[j].X, c[j].Y = 1.0, 1.0
    r[j].X = 4.0 * p[j].X + 2.0 * p[j+1].X
    r[j].Y = 4.0 * p[j].Y + 2.0 * p[j+1].Y
  }

  a[n-1].X, a[n-1].Y = 2.0, 2.0
  b[n-1].X, b[n-1].Y = 7.0, 7.0
  c[n-1].X, c[n-1].Y = 0.0, 0.0
  r[n-1].X = 8.0 * p[n-1].X + p[n].X
  r[n-1].Y = 8.0 * p[n-1].Y + p[n].Y

  for j := 1; j < n; j++ {
    m := a[j].X / b[j-1].X
    b[j].X = b[j].X - m * c[j-1].X
    r[j].X = r[j].X - m * r[j-1].X
    m = a[j].Y / b[j-1].Y
    b[j].Y = b[j].Y - m * c[j-1].Y
    r[j].Y = r[j].Y - m * r[j-1].Y
  }

  // Compute first control point (P1 in BeziefF), from last to first.
  bezier[n-1].P1.X = r[n-1].X / b[n-1].X
  bezier[n-1].P1.Y = r[n-1].Y / b[n-1].Y
  for j := n - 2; j >= 0; j-- {
    bezier[j].P1.X = (r[j].X - c[j].X * bezier[j+1].P1.X) / b[j].X
    bezier[j].P1.Y = (r[j].Y - c[j].Y * bezier[j+1].P1.Y) / b[j].Y
  }

  // Compute second control point (P2 in BezierF), from first to last.
  for j := 0; j < n - 1; j++ {
    bezier[j].P2.X = 2 * p[j].X - bezier[j].P1.X
    bezier[j].P2.Y = 2 * p[j].Y - bezier[j].P1.Y
  }
  bezier[n-1].P2.X = 0.5 * (p[n].X + bezier[n-1].P1.X)
  bezier[n-1].P2.Y = 0.5 * (p[n].Y + bezier[n-1].P1.Y)

  return bezier
}


beziertest/main.go
package main

import (
  "fmt"
  "image"
  "log"
  "os"
  "strconv"
  "strings"
  "github.com/brothertoad/bezier"
)

const prefix = `<?xml version="1.0" standalone="no"?>
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
<svg width="1200px" height="800px" viewBox="0 0 1200 800" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink= "http://www.w3.org/1999/xlink">
  <style>
  * {
    fill: none;
    stroke: black;
    stroke-width: 3;
  }
  .empty {
    fill: none;
    stroke: blue;
    stroke-width: 2;
  }
  .knot {
    fill: blue;
    stroke: blue;
  }
  .control {
    fill: green;
    stroke: green;
  }
  </style>
  <rect x="20" y="20" width="1160" height="760" class="empty"/>
`

const radius = 4

func main() {
  var points []image.Point
  if len(os.Args) > 1 {
    points = argsToPoints(os.Args[1:])
  } else {
    points = argsToPoints([]string{ "200,200", "350,344", "388,431", "460,527", "800,400" })
  }
  // Get the beziers separately so we can plot the control points.
  beziers := bezier.GetControlPointsI(points)
  segments := bezier.ControlPointsToSvgI(beziers)
  fmt.Print(prefix)
  for _, s := range(segments) {
    fmt.Println(`<path d="`, s, `"/>`)
  }
  for _, p := range(points) {
    fmt.Printf(`<circle cx="%d" cy="%d" r="%d" class="knot"/>` + "\n", p.X, p.Y, radius)
  }
  for _, bz := range(beziers) {
    fmt.Printf(`<circle cx="%d" cy="%d" r="%d" class="control"/>` + "\n", bz.P1.X, bz.P1.Y, radius)
    fmt.Printf(`<circle cx="%d" cy="%d" r="%d" class="control"/>` + "\n", bz.P2.X, bz.P2.Y, radius)
  }
  fmt.Println("</svg>")
}

func argsToPoints(args []string) []image.Point {
  points := make([]image.Point, len(args))
  for j, arg := range(args) {
    coords := strings.Split(arg, ",")
    x, err := strconv.Atoi(coords[0])
    if err != nil {
      log.Fatalf("x value %s is not a number\n", coords[0])
    }
    y, err := strconv.Atoi(coords[1])
    if err != nil {
      log.Fatalf("y value %s is not a number\n", coords[1])
    }
    points[j].X = x
    points[j].Y = y
  }
  return points
}


bezier.go
package bezier

import (
  "fmt"
  "image"
)

// The "F" suffix on the struct names means float64.

type PointF struct {
  X, Y float64
}

type BezierF struct {
  P0, P1, P2, P3 PointF
}

type Bezier struct {
  P0, P1, P2, P3 image.Point
}

func SvgControlPointsI(p []image.Point) []string {
  return ControlPointsToSvgI(GetControlPointsI(p))
}

func ControlPointsToSvgI(beziers []Bezier) []string {
  svgs := make([]string, len(beziers))
  for j, b := range(beziers) {
    svgs[j] = "M " + pointAsKnot(b.P0) + " C " + pointAsKnot(b.P1) + ", " +
      pointAsKnot(b.P2) + ", " + pointAsKnot(b.P3)
  }
  return svgs
}

// Converts an image.Point to a string consisting of the x value and
// the y value, separated by a space, suitable for SVG paths.
func pointAsKnot(p image.Point) string {
  return fmt.Sprintf("%d %d", p.X, p.Y)
}

func GetControlPointsI(p []image.Point) []Bezier {
  // Convert integer values to floats, calcalute control points, and then
  // convert back to integers.
  pf := make([]PointF, len(p))
  for j, source := range(p) {
    pf[j].X = float64(source.X)
    pf[j].Y = float64(source.Y)
  }
  bezierF := GetControlPointsF(pf)
  bezier := make([]Bezier, len(bezierF))
  for j, bz := range(bezierF) {
    bezier[j].P0.X = int(bz.P0.X)
    bezier[j].P0.Y = int(bz.P0.Y)
    bezier[j].P1.X = int(bz.P1.X)
    bezier[j].P1.Y = int(bz.P1.Y)
    bezier[j].P2.X = int(bz.P2.X)
    bezier[j].P2.Y = int(bz.P2.Y)
    bezier[j].P3.X = int(bz.P3.X)
    bezier[j].P3.Y = int(bz.P3.Y)
  }
  return bezier
}

// Algorithm ported from https://www.particleincell.com/wp-content/uploads/2012/06/bezier-spline.js
func GetControlPointsF(p []PointF) []BezierF {
  n := len(p) - 1 // number of segments
  bezier := make([]BezierF, n)

  // Fill in the end points of the Bezier curves.
  for j, _ := range(bezier) {
    bezier[j].P0.X, bezier[j].P0.Y = p[j].X, p[j].Y
    bezier[j].P3.X, bezier[j].P3.Y = p[j+1].X, p[j+1].Y
  }

  // Slices needed for calculations, listed as "rhs vector" on page listed above.
  a := make([]PointF, n)
  b := make([]PointF, n)
  c := make([]PointF, n)
  r := make([]PointF, n)

  a[0].X, a[0].Y = 0.0, 0.0
  b[0].X, b[0].Y = 2.0, 2.0
  c[0].X, c[0].Y = 1.0, 1.0
  r[0].X = p[0].X + 2.0 * p[1].X
  r[0].Y = p[0].Y + 2.0 * p[1].Y

  for j := 1; j < (n-1); j++ {
    a[j].X, a[j].Y = 1.0, 1.0
    b[j].X, b[j].Y = 4.0, 4.0
    c[j].X, c[j].Y = 1.0, 1.0
    r[j].X = 4.0 * p[j].X + 2.0 * p[j+1].X
    r[j].Y = 4.0 * p[j].Y + 2.0 * p[j+1].Y
  }

  a[n-1].X, a[n-1].Y = 2.0, 2.0
  b[n-1].X, b[n-1].Y = 7.0, 7.0
  c[n-1].X, c[n-1].Y = 0.0, 0.0
  r[n-1].X = 8.0 * p[n-1].X + p[n].X
  r[n-1].Y = 8.0 * p[n-1].Y + p[n].Y

  for j := 1; j < n; j++ {
    m := a[j].X / b[j-1].X
    b[j].X = b[j].X - m * c[j-1].X
    r[j].X = r[j].X - m * r[j-1].X
    m = a[j].Y / b[j-1].Y
    b[j].Y = b[j].Y - m * c[j-1].Y
    r[j].Y = r[j].Y - m * r[j-1].Y
  }

  // Compute first control point (P1 in BeziefF), from last to first.
  bezier[n-1].P1.X = r[n-1].X / b[n-1].X
  bezier[n-1].P1.Y = r[n-1].Y / b[n-1].Y
  for j := n - 2; j >= 0; j-- {
    bezier[j].P1.X = (r[j].X - c[j].X * bezier[j+1].P1.X) / b[j].X
    bezier[j].P1.Y = (r[j].Y - c[j].Y * bezier[j+1].P1.Y) / b[j].Y
  }

  // Compute second control point (P2 in BezierF), from first to last.
  for j := 0; j < n - 1; j++ {
    bezier[j].P2.X = 2 * p[j+1].X - bezier[j+1].P1.X
    bezier[j].P2.Y = 2 * p[j+1].Y - bezier[j+1].P1.Y
  }
  bezier[n-1].P2.X = 0.5 * (p[n].X + bezier[n-1].P1.X)
  bezier[n-1].P2.Y = 0.5 * (p[n].Y + bezier[n-1].P1.Y)

  return bezier
}
